### ISSUE TITLE
Example using Meteor Method

### ISSUE THREAD

**Issue by tdmoneybanks** [vsivsi/meteor-job-collection#71]
Hi, any chance you could do an examle/gist of using creating a job and returning the result to the user within a meteor method? Great job on the project!

**Comment by vsivsi** [comment]
Hi, can you be a bit more specific about what you'd like to do?

The are a bunch of ways that this could be done, but the best way will depend on what you are trying to accomplish.

If you are totally new to job-colleciton, you should definitely check out the [sample app](https://github.com/vsivsi/meteor-file-job-sample-app).

**Comment by tdmoneybanks** [comment]
Hey,
so more specifically, I would like to use a meteor method that creates and
saves the job (that gets pushed to a processJobs queue). This job would
then be used to create a user account after grabbing some data from a
remote server. It would then be marked as completed and the meteor method
would handle logging in the user and sending the user to the new route for
logging in.

**Comment by rhyslbw** [comment]
@tdmoneybanks 

Sounds like you need to setup an [observer-based worker group](https://github.com/vsivsi/meteor-job-collection#externally-trigger-the-jobqueue-to-seek-new-work), then just create a job on the client as per usual by instantiating the job class:

    // Create a job:
    var job = new Job(myJobs, 'sendEmail', // type of job
      // Job data that you define, including anything the job
      // needs to complete. May contain links to files, etc...
      {
        address: 'bozo@clowns.com',
        subject: 'Critical rainbow hair shortage',
        message: 'LOL; JK, KThxBye.'
      }
    );

    // Set some properties of the job and then submit it
    job.priority('normal')
      .retry({ retries: 5,
        wait: 15*60*1000 })  // 15 minutes between attempts
      .delay(60*60*1000)     // Wait an hour before first try
      .save();               // Commit it to the server

Check the [quick example](https://github.com/vsivsi/meteor-job-collection#quick-example)

**Comment by vsivsi** [comment]
@tdmoneybanks  Check if I understand the flow of this...
1.  Client calls Meteor method on Meteor server requesting a new account
2.  Server method code creates a job to set-up new account, And waits for the job to complete
3.  Another server runs the job and somehow communicates back to the Meteor server that it is complete.
4.  Waiting method sees results of completed job and completes login using new account

If I have that correct, then there are two things you need to decide:

*  Do you really want to insert a job queue into the process of creating a new account while a user is waiting?  The nature of job queues is that jobs may not run immediately. There may be no workers immediately available for any number of reasons, and in any case it will add latency. I may still not know enough about the exact situation, but this feels more like a situation when RPC is a better fit. Is the second server also a Meteor server?  If not, is there some other remote API that can be used to call it directly, or is there some reason you want it to be pulling work out of a job queue?

*  Assuming a job queue is still the right solution for this case in your application, then you need to decide how the remote job will communicate its results to the waiting Meteor method call. Probably the easiest way would be via a shared document in MongoDB. So something like:
    * Server method creates a new placeholder document in the collection, and adds the `_id` value of the document to the `data` object of the new job it creates.  It's possible that this shared document could be the job document itself, in which case the `_id` is built-in.
    * Remote server worker completes its work and writes results to the MongoDB document corresponding to `_id`. If the job document itself is the shared document, then the results can be shared back through the call to `job.done()`
    * Server method is waiting on a change to the placeholder document (using [`cursor.observe()`](http://docs.meteor.com/#/full/observe), validates the new user data and then returns.

As I said in my initial response, there are a bunch of ways this could all be done. For example the decision whether to include the remote worker results in a separate document or within the results of the job itself will depend on the nature of that information and whether you want it living in your job-queue.  My guess is that your wouldn't want that (old jobs get deleted...) so that's why I brought up the "other document" idea.  

It's also possible that the remote worker creates the new account document itself, and just shares back its `_id` in the job results. The method code would then wait on the job to complete and then grab the `_id` from the results and do a separate lookup to validate and login the new user.

Anyway, as I said initially, there are a bunch of ways you could do this.

**Comment by tdmoneybanks** [comment]
Yeah I was not clear enough in my original explanation so let me try again! Also, I wasn't exactly sure on the best way to do this in meteor as I tried to do my own reactive queue and didn't have great results. So, I only want to create users if they have an account with another site. I am using an api and meteor's http to call the remote server and it is returning JSON data about the user if they have an account there. The api has a rate limit call so I thought using jobs would be a decent method since they can retry jobs if they fail due to the rate limit being hit note* I also considered just grabbing the timeout rate and using that as the param for a meteor.setTimeout() call. So here are the steps as I see them:
1. user fills in form and triggers meteor call
2. meteor call creates a job and passes the users input data to the job.
3. job hits remote api and checks data.
4. job creates account and pushes data into the account then calls done()
5. client side checks for job status and sends them to the new route when its finished
6. the loading screen for users (triggered by the event and using job data for progress) disappears with the new route.

I may still be needed a different solution though as this is my first meteor app. Please give me your input and let me know! Thanks

**Comment by vsivsi** [comment]
Hi, yes job-collection could certainly do this, but it seems like a bit of overkill for this application, honestly.  The setTimeout solution may work fine, or something a bit more powerful, like [powerqueue](https://atmospherejs.com/cfs/power-queue) may be more appropriate.  Job-collection does have some nice features like exponential back off for rate limiting, but again it seems a bit heavy for this pretty straightforward case. 

The flow you have above should work however. He main thing is that at step 5, the client must subscribe to a published cursor on the job collection to get that notification. Obviously that server publish function need to be very locked down to prevent an attacker from being able to snoop on these authentication jobs. As @rhyslbw mentioned above, you can give the client permission to create the job itself and dispense with the separate meteor call.  So in that case:

1.  Client creates new job with account info in `data`.  Waits on job completion using a reactive computation on the job or observe in the cursor. 
2. Job runs on the server, queries remote api, when successful. Calls job.done() will credentials
3. Client reactively responds to successful job, proceeds with login. 

His would all work fine, but as I said above, you would really need to lock down the job-collection pub/sub to ensure that each client could only see its own job, and the job-collection allow/deny rules to only allow a client to create a job with very specific properties, and deny all other actions. 

Anyway, welcome to the world of Meteor!  I hope this helps.

**Comment by tdmoneybanks** [comment]
Thanks so much for the response! The job does not need any sensitive information to hit the remote api. Is there a better way to do this functionality without giving the job that info? Like, create the job without the pass/email and have it just return the remote data and then create the account client side?

**Comment by vsivsi** [comment]
It's sounds like the Meteor server needs to receive the account info somehow.  Whether you make a custom Meteor method to pass it, or you create a job on the client to pass it (which is just a special kind of Meteor method call...). It's really up to you how you want to structure this.

Account security is hard, and I don't feel like I can give you advice on how you should design this to work, given that I don't know any of the necessary details, and because this is not really a job-collection issue, per se.

**Comment by tdmoneybanks** [comment]
completely understand. Thanks again for the help you can close this issue!

**Comment by vsivsi** [comment]
No worries, and good luck!
